
Starting with commit 'backend-drm: Use the seat name over cmd line' we
would be able to create a fake-seat, by just specifying an invalid
seat name.

In the same time, in a system where there are actually no input devices
and no one specifies an explicit seat (as in to create a fake one) we
would end with no seat available in the system.

This makes things a bit easier for situations where users would just
specify a DRM node, with input devices being accessed directly, or
rather with no input devices whatsoever.
Signed-off-by: Marius Vlad's avatarMarius Vlad <marius.vlad@collabora.com>

from https://gitlab.freedesktop.org/wayland/weston/-/merge_requests/798/diffs?commit_id=1639b9c6995185383f13d0e2e26fee06b7de5dde


diff --speed-large-files --no-dereference --minimal -Naur weston-10.0.1/libweston/backend-drm/drm.c weston-10.0.1/libweston/backend-drm/drm.c
--- weston-10.0.1/libweston/backend-drm/drm.c	2022-07-22 12:26:58.963329289 +0200
+++ weston-10.0.1/libweston/backend-drm/drm.c	2022-07-22 12:26:43.869996049 +0200
@@ -1230,22 +1230,25 @@
 			     struct weston_output *output,
 			     const char *s)
 {
-	if (strcmp(s, "") != 0) {
-		struct weston_pointer *pointer;
-		struct udev_seat *seat;
+	struct weston_pointer *pointer;
+	struct udev_seat *seat;
+	const char *seat_name = NULL;
 
-		seat = udev_seat_get_named(&b->input, s);
-		if (!seat)
-			return;
+	seat_name = s;
+	if (strcmp(seat_name, "") == 0)
+		seat_name = "fake-seat";
 
-		seat->base.output = output;
+	seat = udev_seat_get_named(&b->input, seat_name);
+	if (!seat)
+		return;
 
-		pointer = weston_seat_get_pointer(&seat->base);
-		if (pointer)
-			weston_pointer_clamp(pointer,
-					     &pointer->x,
-					     &pointer->y);
-	}
+	seat->base.output = output;
+
+	pointer = weston_seat_get_pointer(&seat->base);
+	if (pointer)
+		weston_pointer_clamp(pointer,
+				&pointer->x,
+				&pointer->y);
 }
 
 static int
@@ -1362,6 +1365,9 @@
 	struct drm_output *output = to_drm_output(base);
 	struct drm_backend *b = to_drm_backend(base->compositor);
 
+	if (b->seat_id)
+		seat = b->seat_id;
+
 	setup_output_seat_constraint(b, &output->base,
 				     seat ? seat : "");
 }
@@ -2592,6 +2598,7 @@
 	weston_launcher_destroy(ec->launcher);
 
 	free(b->drm.filename);
+	free(b->seat_id);
 	free(b);
 }
 
@@ -3009,6 +3016,8 @@
 	b->use_pixman = config->use_pixman;
 	b->pageflip_timeout = config->pageflip_timeout;
 	b->use_pixman_shadow = config->use_pixman_shadow;
+	if (config->seat_id)
+		b->seat_id = strdup(config->seat_id);
 
 	b->debug = weston_compositor_add_log_scope(compositor, "drm-backend",
 						   "Debug messages from DRM/KMS backend\n",
@@ -3222,6 +3231,7 @@
 	if (b->gbm)
 		gbm_device_destroy(b->gbm);
 #endif
+	free(b->seat_id);
 	free(b);
 	return NULL;
 }
diff --speed-large-files --no-dereference --minimal -Naur weston-10.0.1/libweston/backend-drm/drm-internal.h weston-10.0.1/libweston/backend-drm/drm-internal.h
--- weston-10.0.1/libweston/backend-drm/drm-internal.h	2022-06-23 15:52:38.000000000 +0200
+++ weston-10.0.1/libweston/backend-drm/drm-internal.h	2022-07-22 12:24:32.363330158 +0200
@@ -296,6 +296,7 @@
 
 	bool use_pixman;
 	bool use_pixman_shadow;
+	char *seat_id;
 
 	struct udev_input input;
 
